# Analysis to determine a Trend in Hacker News 


A dataset of the posts on hackerank would be analyzed to determine which sort of post recieve more comment. And also, at what time of the day is one likely to recieve the most comment on a post



In [0]:
import csv

#reading in the csv file and displying the first 5 rows
file = open('hacker_news.csv')
file_read = csv.reader(file)
hn = list(file_read)

print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [0]:
#removing the column headers
headers = hn[0]

hn = hn[1:]
hn[:5]

[['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12'],
 ['10482257',
  'Title II kills investment? Comcast and other ISPs are now spending more',
  'http://arstechnica.com/business/2015/10/comcast-and-other-isps-boost-network-

In [0]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print('Ask Posts are: ', len(ask_posts))
print('\n')
print('Show Posts are: ', len(show_posts))
print('\n')
print('Other Posts are: ', len(other_posts))
    

Ask Posts are:  1744


Show Posts are:  1162


Other Posts are:  17193



In the cell above, the hn dataset was sorted into three lists depending on the string that starts the title parameter. The ask post after the sorting are 1744 in number while the show post are 1162. Other post totaled 17193. 


In [0]:
total_ask_comments = 0
total_show_comments = 0

def comments(post):
    total_comments = 0
    for row in post:
        total_comments += int(row[4])
    return total_comments

total_ask_comments = comments(ask_posts)
total_show_comments = comments(show_posts)

avg_ask_comments = total_ask_comments / len(ask_posts)
avg_show_comments = total_show_comments / len(show_posts)

print('Average Ask Post is: ', avg_ask_comments)
print('\n')
print('Average Show Post is: ', avg_show_comments)

Average Ask Post is:  14.038417431192661


Average Show Post is:  10.31669535283993


In the cell above, it was shown that averagely, a post asking a question would recieve more comments under it than a post showing a certain finding on Hackerank. 

In [0]:
import datetime as dt

result_list = []

for row in ask_posts:
    result_list.append([row[6], int(row[4])])    

counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    dt_obj = dt.datetime.strptime(row[0], '%m/%d/%Y %H:%M').strftime('%H')
    if dt_obj not in counts_by_hour:
        counts_by_hour[dt_obj] = 1
        comments_by_hour[dt_obj] = row[1]
    else:
        counts_by_hour[dt_obj] += 1
        comments_by_hour[dt_obj] += row[1]

In the cell above, we created two dictionaries whose key are the hours of post creation. The counts_by_hour dictionary contains the number of post in a particular hour while the comments_by_hour contains the corresponding number of comments.

In [0]:
avg_by_hour = []

for key in counts_by_hour:
    
    avg_by_hour.append([key ,comments_by_hour[key] / counts_by_hour[key]])

avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

In [0]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
sorted_swap = sorted(swap_avg_by_hour, reverse=True)

print("Top 5 Hours for Ask Posts Comments")
for row in sorted_swap[:5]:
    dt_obj = dt.datetime.strptime(row[1], '%H').strftime('%H:00')
    print('{} : {:.2f} average comments per post'.format(dt_obj, row[0]))

Top 5 Hours for Ask Posts Comments
15:00 : 38.59 average comments per post
02:00 : 23.81 average comments per post
20:00 : 21.52 average comments per post
16:00 : 16.80 average comments per post
21:00 : 16.01 average comments per post


The result above presents the top 5 hour of the day one is likely to receive the most comment on an Ask post on Hackerank. However, the time zone of the dataset is US Eastern time. Therefore, it is evident that to get the most comment on the post, is should be around 15:00 UTC, 2:00 UTC or perhaps 20:00 UTC 